src/main/java/net/snowflake/ingest/streaming/internal/ParquetFlusher.java

```diff
@@ -0,0 +1,271 @@
+    this.schema = schema;
+  }
+
+  @Override
+  public SerializationResult serialize(
+      List<ChannelData<ParquetRowBuffer.ParquetChunkData>> channelsDataPerTable,
+      ByteArrayOutputStream chunkData,
+      String filePath)
+      throws IOException {
+    List<ChannelMetadata> channelsMetadataList = new ArrayList<>();
+    long rowCount = 0L;
+    SnowflakeStreamingIngestChannelInternal<ParquetRowBuffer.ParquetChunkData> firstChannel = null;
+    Map<String, RowBufferStats> columnEpStatsMapCombined = null;
+    List<List<Object>> rows = null;
+
+    for (ChannelData<ParquetRowBuffer.ParquetChunkData> data : channelsDataPerTable) {
+      // Create channel metadata
+      ChannelMetadata channelMetadata =
+          ChannelMetadata.builder()
+              .setOwningChannel(data.getChannel())
+              .setRowSequencer(data.getRowSequencer())
+              .setOffsetToken(data.getOffsetToken())
+              .build();
+      // Add channel metadata to the metadata list
+      channelsMetadataList.add(channelMetadata);
+
+      logger.logDebug(
+          "Parquet Flusher: Start building channel={}, rowCount={}, bufferSize={} in blob={}",
+          data.getChannel().getFullyQualifiedName(),
+          data.getRowCount(),
+          data.getBufferSize(),
+          filePath);
+
+      if (rows == null) {
+        columnEpStatsMapCombined = data.getColumnEps();
+        rows = new ArrayList<>();
+        firstChannel = data.getChannel();
+      } else {
+        // This method assumes that channelsDataPerTable is grouped by table. We double check
+        // here and throw an error if the assumption is violated
+        if (!data.getChannel()
+            .getFullyQualifiedTableName()
+            .equals(firstChannel.getFullyQualifiedTableName())) {
+          throw new SFException(ErrorCode.INVALID_DATA_IN_CHUNK);
+        }
+
+        columnEpStatsMapCombined =
+            ChannelData.getCombinedColumnStatsMap(columnEpStatsMapCombined, data.getColumnEps());
+      }
+      rows.addAll(data.getVectors().rows);
+
+      rowCount += data.getRowCount();
+
+      logger.logDebug(
+          "Parquet Flusher: Finish building channel={}, rowCount={}, bufferSize={} in blob={}",
+          data.getChannel().getFullyQualifiedName(),
+          data.getRowCount(),
+          data.getBufferSize(),
+          filePath);
+    }
+
+    Map<String, String> metadata = channelsDataPerTable.get(0).getVectors().metadata;
+
+    flushToParquet(chunkData, rows, metadata);
+    return new SerializationResult(channelsMetadataList, columnEpStatsMapCombined, rowCount);
+  }
+
+  private void flushToParquet(
+      ByteArrayOutputStream output, List<List<Object>> rows, Map<String, String> metadata)
+      throws IOException {
+    try {
+      ParquetWriter<List<Object>> writer =
+          new BdecParquetWriterBuilder(output, schema, metadata)
+              .withWriterVersion(ParquetProperties.WriterVersion.PARQUET_1_0)
```

Comment:
Why aren't we using the latest version `WriterVersion.PARQUET_2_0`? I would assume that `1_0` is used server-side and we have to match that.

Reply:
PARQUET_2_0 uses `Encoding.DELTA_BYTE_ARRAY` for byte arrays (e.g. SF sb16)
Server side scanner does not support it yet, this may be an easy fix in future